storage.go

```diff
@@ -73,61 +80,55 @@ func (s *Storage) initConfig() error {
 	if s.LockPollingInterval == 0 {
 		s.LockPollingInterval = lockPollingInterval
 	}
+	if isEmpty(s.KeyListDataKey[:]) {
+		s.KeyListDataKey = crypto.HashBytes([]byte("key_list"))
+	}
 	return nil
 }
 
 // Store puts value at key.
 func (s *Storage) Store(key string, value []byte) error {
-	if err := s.initConfig(); err != nil {
+	err := s.initConfig()
+	if err != nil {
 		return err
 	}
 
 	if key == "" {
 		return errors.New("key must not be empty")
 	}
-	// Get the item.
-	it, rev, err := s.getItem(key)
-	if err != nil && !errors.Contains(err, errNotExist) {
-		return err
-	}
-	keyList, keyListRev, err := s.keyList()
-	if err != nil && !errors.Contains(err, skydb.ErrNotFound) {
-		return err
-	}
-	if keyList == nil {
-		keyList = make(map[string]bool)
-	}
-
-	keyListChanged := false
-	// Are we deleting the entry? If so, remove the key from the key list.
-	// Otherwise add it to the list.
-	if slicesEqual(value, emptyRegistryEntry[:]) {
-		delete(keyList, key)
-		keyListChanged = true
-	} else {
-		_, exists := keyList[key]
-		if !exists {
-			keyList[key] = true
-			keyListChanged = true
+	// Are we deleting the entry? If we're not deleting item then we're adding item.
+	isDeletion := slicesEqual(value, emptyRegistryEntry[:])
+	// Modifying the keylist might fail because somebody else has locked item.
+	// We will retry several times.
+	for triesLeft := keylistModificationAttempts; triesLeft > 0; triesLeft-- {
+		if isDeletion {
+			err = s.keyListDelete(key)
+		} else {
+			err = s.keyListAdd(key)
 		}
-	}
-	if keyListChanged {
-		bytes, err := json.Marshal(keyList)
-		if err != nil {
-			return errors.AddContext(err, "failed to serialise a new key list")
+		if err == nil {
+			break
 		}
-		err = s.SkyDB.Write(bytes, s.KeyListDataKey, keyListRev+1)
-		if err != nil {
-			return errors.AddContext(err, "failed to store the key list")
+		if triesLeft > 0 {
+			fmt.Println("failed to modify keylist, will try again. error:", err)
 		}
+		time.Sleep(keylistModificationSleep * time.Second)
+	}
+	if err != nil {
+		return errors.AddContext(err, "failed to modify keylist")
 	}
 
+	// Get the item.
+	item, rev, err := s.getItem(key)
+	if err != nil && !errors.Contains(err, errNotExist) {
+		return err
+	}
 	// Update the item.
-	it.PrimaryKey = key
-	it.Contents = value
-	it.LastUpdated = time.Now().UTC()
+	item.PrimaryKey = key
+	item.Contents = value
+	item.LastUpdated = time.Now().UTC()
 	// Store the key's new value
-	bytes, err := json.Marshal(it)
+	bytes, err := json.Marshal(item)
```

Comment:
Do we always `json.Marshal` before writing? If so we might want to deduplicate this a bit with a helper that fetches the item first, updates it, marshals it and writes it to skydb.

Reply:
Good point, we can dedup this.

The fetching of the item is not really part of the broad model, it's more of a local use case. So, the helper constructs the item, marshals it, derives the dataKey and stores it in SkyDB.